src/psyGen.py

```diff
@@ -1930,43 +2266,74 @@ def call(self):
-        self._anti_dependence = []
-        for following_call in self._call.following_calls:
-            for argument in following_call.arguments.args:
-                if argument.name == self._name:
-                    if self.access in writers and argument.access in readers:
-                        self._true_dependence.append(argument)
-                    if self.access in readers and argument.access in writers:
-                        self._anti_dependence.append(argument)
-
-    def has_true_dependence(self):
-        if len(self._true_dependence) > 0:
-            return True
-        return False
-
-    def has_anti_dependence(self):
-        if len(self._anti_dependence) > 0:
-            return True
-        return False
+    @call.setter
+    def call(self, value):
+        ''' set the node that this argument is associated with '''
+        self._call = value
+
+    def backward_dependence(self):
+        '''Returns the preceding argument that this argument has a direct
+        dependence with, or None if there is not one. The argument may
+        exist in a call, a haloexchange, or a globalsum. For performance
+        reasons only compute once then store the result. '''
+        if not self._bd_computed:
+            self._bd_computed = True
+            all_nodes = self._call.walk(self._call.root.children, Node)
+            position = all_nodes.index(self._call)
+            all_prev_nodes = all_nodes[:position]
+            all_prev_nodes.reverse()
+            self._bd_value = self._find_argument(all_prev_nodes)
+        return self._bd_value
+
+    def forward_dependence(self):
+        '''Returns the following argument that this argument has a direct
+        dependence with, or None if there is not one. The argument may
+        exist in a call, a haloexchange, or a globalsum. For performance
+        reasons only compute once, then store the result.'''
+        if not self._fd_computed:
+            self._fd_computed = True
+            all_nodes = self._call.walk(self._call.root.children, Node)
+            position = all_nodes.index(self._call)
+            all_following_nodes = all_nodes[position+1:]
+            self._fd_value = self._find_argument(all_following_nodes)
+        return self._fd_value
+
+    def _find_argument(self, nodes):
+        '''Return the first argument in the list of nodes that has a
+        dependency with self. If one is not found return None'''
+        for node in nodes:
+            if isinstance(node, Call):
+                for argument in node.arguments.args:
+                    if self._dependent_arg(argument):
+                        return argument
+            elif isinstance(node, HaloExchange):
+                argument = node.field
+                if self._dependent_arg(argument):
+                    return argument
+            elif isinstance(node, GlobalSum):
+                argument = node.scalar
+                if self._dependent_arg(argument):
```

Comment:
Could use the new `args` method of `GlobalSum` and `HaloExchange` here. e.g.

    elif isinstance(node, HaloExchange) or isinstance(node, GlobalSum):
        for argument in node.args:
            if self._dependent_arg(argument):
                return argument